### PR TITLE
Towards fullgeneric 5

### PR DESCRIPTION
**Pull request by cseed**
Builds on: https://github.com/hail-is/hail/pull/2074

Added optimized unsafe row add to region value builder.  Tests are faster than toward_fullgeneric_4, 0.1 (8m2s vs 9m18s, 0.1: 8m20s).

### CODE REVIEW

File: src/main/scala/is/hail/expr/Type.scala [hunk 1]
```diff
@@ -191,6 +191,8 @@ sealed abstract class Type extends Serializable { self =>
   /*  Fundamental types are types that can be handled natively by RegionValueBuilder: primitive
       types, Array and Struct. */
   lazy val fundamentalType: Type = this
+
+  def containsPointer: Boolean = children.exists(_.containsPointer)
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/Type.scala]: This seems wrong. I think it should be `fundamentalType.containsPointer || fundamentalType.children.exists(_.containsPointer)`

File: src/test/scala/is/hail/SparkSuite.scala [hunk 2]
```diff
@@ -9,7 +9,7 @@ import org.scalatest.testng.TestNGSuite
 object SparkSuite {
   lazy val hc = HailContext(master = Option(System.getProperty("hail.master")),
     appName = "Hail.TestNG",
-    local = "local[2]",
```
Comment (tpoterba) [inline src/test/scala/is/hail/SparkSuite.scala]: intentional change? Maybe this is why the tests are faster...

File: src/main/scala/is/hail/annotations/MemoryBlock.scala [hunk 3]
```diff
@@ -409,11 +401,118 @@ class RegionValueBuilder(region: MemoryBuffer) {
...
+    val length = fromRegion.loadInt(fromBOff)
+
+    region.align(4)
+    val toBOff = region.allocate(4 + length)
+    region.storeInt(toOff, toBOff)
+
+    region.copyFrom(fromRegion, fromBOff, toBOff, 4 + length)
+  }
+
+  def requiresFixup(t: Type): Boolean = {
```
Comment (cseed) [inline src/main/scala/is/hail/annotations/MemoryBlock.scala]: Yes.  See the entrypoint into the fixup routines, like `addUnsafeRow`, which calls `fixupStruct`:

```
    fixupStruct(t.fundamentalType, toOff, fromRegion, fromOff)
```

Hmm, yeah, I think I like that.  I'll make that change.

File: src/test/scala/is/hail/SparkSuite.scala (hunk 2 above)
Comment (cseed) [inline src/test/scala/is/hail/SparkSuite.scala]: Oh, shit, yes.  I'll rerun.

File: src/main/scala/is/hail/expr/Type.scala [hunk 4]
```diff
@@ -730,7 +743,7 @@ case class TDict(keyType: Type, valueType: Type) extends TContainer {
 case object TGenotype extends ComplexType {
   override def toString = "Genotype"
 
-  override lazy val fundamentalType: TStruct = TStruct(
+  override lazy val representation: TStruct = TStruct(
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/Type.scala]: why go backwards here?

Comment (cseed) [inline src/main/scala/is/hail/expr/Type.scala]: `representation` is used to build the Spark schema when converting to a data frame, e.g. `kt.toDF`, but there we need to keep the distinction between Binary and String.

Comment (tpoterba) [inline src/main/scala/is/hail/expr/Type.scala]: ahhh, got it

**Pull request merged**